File: src/solutions/python/SetMatrixZeroes.py
from typing import List

'''
# Problem
	- `Link`: https://leetcode.com/problems/set-matrix-zeroes/
# Solution
	- `Author`: Kyungtaek Lim (Jonas)
	- `Date`: Apr 10, 2025
	- `Answer`: setZeroes / setZeroesAdvanced
'''
class Solution:

    '''
    # Option #1
    - Common Way
    - O (m * n)
    - Space Complexity: O(m + n)
    '''
    def setZeroes(self, matrix: List[List[int]]) -> None:
        """
        Do not return anything, modify matrix in-place instead.
        """
        m = len(matrix)
        n = len(matrix[0])

        row_set = set()
        col_set = set()
        
        for i in range(m):
            for j in range(n):
                if not matrix[i][j]:
                    row_set.add(i)
                    col_set.add(j)
        
        for i in range(m):
            for j in range(n):
                if i in row_set or j in col_set:
                    matrix[i][j] = 0


    '''
    # Option #2
    - Advanced Way
    - O (m * n)
    - Space Complexity: O(1)
    '''
    def setZeroesAdvanced(self, matrix: List[List[int]]) -> None:
        """
        Do not return anything, modify matrix in-place instead.
        """
        m, n = len(matrix), len(matrix[0])
        row, col = False, False

        # Check First row & col
        for j in range(n):
            if not matrix[0][j]:
                row = True
                break
        for i in range(m):
            if not matrix[i][0]:
                col = True
                break
        
        # Check The other row
        for i in range(1, m):
            for j in range(1, n):
                if not matrix[i][j]:
                    matrix[0][j] = 0
                    matrix[i][0] = 0
        
        # Put 0 based on the first row & col
        for j in range(1, n):
            if not matrix[0][j]:
                for i in range(1, m):
                    matrix[i][j] = 0

        for i in range(1, m):
            if not matrix[i][0]:
                for j in range(1, n):
                    matrix[i][j] = 0
        
        # Change First row & col
        if row:
            for j in range(n):
                matrix[0][j] = 0

        if col:
            for i in range(m):
                matrix[i][0] = 0

File: src/solutions/java/SetMatrixZeroes.java
package solutions.java;

import java.util.HashSet;
import java.util.Set;

/*
# Problem
	- `Link`: https://leetcode.com/problems/set-matrix-zeroes/
# Solution
	- `Author`: Kyungtaek Lim (Jonas)
	- `Date`: Apr 10, 2025
	- `Answer`: setZeroes / setZeroesAdvanced
 */
public class SetMatrixZeroes {
	
	/*
    # Option #1
    - Common Way
    - O (m * n)
    - Space Complexity: O(m + n)
	 */
    public void setZeroes(int[][] matrix) {
        
    	int m = matrix.length, n = matrix[0].length;
    	
    	Set<Integer> rowSet = new HashSet<>();
    	Set<Integer> colSet = new HashSet<>();
    	
    	for (int i = 0; i < m; i++) {
    		for (int j = 0; j < n; j++) {
    			if (matrix[i][j] == 0) {
    				rowSet.add(i);
    				colSet.add(j);
    			}
    		}
    	}
    	
    	for (int i = 0; i < m; i++) {
    		for (int j = 0; j < n; j++) {
    			if (rowSet.contains(i) || colSet.contains(j)) {
    				matrix[i][j] = 0;
    			}
    		}
    	}
    }
    
    /*
    # Option #2
    - Advanced Way
    - O (m * n)
    - Space Complexity: O(1)
     */
    public void setZeroesAdvanced(int[][] matrix) {
        
        final int m = matrix.length, n = matrix[0].length;

        // Check first row & col
        boolean row = false, col = false;
        for (int j = 0; j < n; j++) {
            if (matrix[0][j] == 0) {
                row = true;
                break;
            }
        }
        for (int i = 0; i < m; i++) {
            if (matrix[i][0] == 0) {
                col = true;
                break;
            }
        }

        // Put 0 on the first row & col
        for (int i = 1; i < m; i++) {
            for (int j = 1; j < n; j++) {
                if (matrix[i][j] == 0) {
                    matrix[0][j] = 0;
                    matrix[i][0] = 0;
                }
            }
        }

        // Replace value based on the first row & col
        for (int j = 1; j < n; j++) {
            if (matrix[0][j] == 0) {
                for (int i = 1; i < m; i++){
                    matrix[i][j] = 0;
                }
            }
        }
        for (int i = 1; i < m; i++) {
            if (matrix[i][0] == 0) {
                for (int j = 1; j < n; j++) {
                    matrix[i][j] = 0;
                }
            }
        }

        // Replace values on the first row & col
        if (row) {
            for (int j = 0; j < n; j++) {
                matrix[0][j] = 0;
            }
        }
        if (col) {
            for (int i = 0; i < m; i++) {
                matrix[i][0] = 0;
            }
        }
    }
}
